realm/src/main/java/io/realm/BaseRealm.java

```diff
@@ -586,14 +626,278 @@ public static synchronized void migrateRealm(RealmConfiguration configuration, R
     private class RealmCallback implements Handler.Callback {
         @Override
         public boolean handleMessage(Message message) {
-            if (message.what == REALM_CHANGED) {
-                sharedGroupManager.advanceRead();
-                sendNotifications();
+            switch (message.what) {
+                case REALM_CHANGED: {
+                    if (threadContainsAsyncQueries()) {
```

Comment:
Each of these switch-statements are really long which makes it difficult to get a sense of the control structure in `handleMessage`. Each switch case should be refactored into it's own private method.